Use COPY when executing distributed INSERTs

erimatnor opened the pull request:
A new custom plan/executor node is added that implements distributed
INSERT using COPY in the backend (between access node and data
nodes). COPY is significantly faster than the existing method that
sets up prepared INSERT statements on each data node. With COPY,
tuples are streamed to data nodes instead of batching them in order to
"fill" a configured prepared statement. A COPY also avoids the
overhead of having to plan the statement on each data node.

Using COPY doesn't work in all situations, however. Neither ON
CONFLICT nor RETURNING clauses work since COPY lacks support for
them. Still, RETURNING is possible if one knows that the tuples aren't
going to be modified by, e.g., a trigger. When tuples aren't modified,
one can return the original tuples on the access node.

In order to implement the new custom node, some refactoring has been
performed to the distributed COPY code. The basic COPY support
functions have been moved to the connection module so that switching
in and out of COPY_IN mode is part of the core connection
handling. This allows other parts of the code to manage the connection
mode, which is necessary when, e.g., creating a remote chunk. To
create a chunk, the connection needs to be switched out of COPY_IN
mode so that regular SQL statements can be executed again.

Partial fix for #3025.

erimatnor commented:
Some simple benchmarks for this change below. The runs are as follows:

1. Insert into distributed hypertable with _new_ plan/executor node (`DataNodeCopy`). ~40s
2. Insert with _old_ `DataNodeDispatch` plan/executor node. ~2m53s
3. Insert into regular hypertable (non-distributed) ~53s

All tests running on a single machine and instance of PG.

```
nyc_data=# INSERT INTO rides SELECT * FROM rides_temp LIMIT 1000000;
INSERT 0 1000000
Time: 40254,472 ms (00:40,254)
nyc_data=# set timescaledb.enable_distributed_insert_with_copy=false;
SET
Time: 0,928 ms
nyc_data=# TRUNCATE rides;
TRUNCATE TABLE
Time: 226,613 ms
nyc_data=# INSERT INTO rides SELECT * FROM rides_temp LIMIT 1000000;
INSERT 0 1000000
Time: 173769,384 ms (02:53,769)
nyc_data=# TRUNCATE rides_single;
TRUNCATE TABLE
Time: 20,854 ms
nyc_data=# INSERT INTO rides_single SELECT * FROM rides_temp LIMIT 1000000;
INSERT 0 1000000
Time: 53182,156 ms (00:53,182)
```

nikkhils commented on tsl/src/nodes/data_node_copy.c:
```diff
@@ -0,0 +1,414 @@
+ * separate between native PostgreSQL types with "fixed" Oids,
+ * vs. user-created types with generated Oids.
+ */
+static bool type_is_custom(Oid typeid) { return type_is_rowtype(typeid); }
```
I had suggested `type_is_rowtype(typeid) && typeid > MyDatabaseId`. Doesn't this correctly ensure that the type is a user created custom datatype? Anyways, it's not a big deal as discussed earlier.

erimatnor replied:
I don't think so. rowtypes only cover composite types and not other custom types (e.g., varlen types implemented in C). 

I didn't understand the typeid > MyDatabaseId comparison. Why would it make sense to compare a type OID with a Database OID?

Admittedly, this type checking is imperfect, but I don't think there's a consistent/reliable way to know that a type is user-created.

nikkhils replied:
The database has to be created FIRST with a lower OID, any new objects in this database will have OID more than the dboid was my point.

nikkhils replied:
Plus for this use of `FirstNormalObjectId`

nikkhils commented on tsl/src/planner.c:
```diff
@@ -146,3 +149,82 @@ tsl_set_rel_pathlist(PlannerInfo *root, RelOptInfo *rel, Index rti, RangeTblEntr
+				 * Check for BEFORE INSERT triggers as those are the ones that can
+				 * modify a tuple.
+				 */
+				if (strcmp(trig->tgname, INSERT_BLOCKER_NAME) != 0 &&
```
So trigger checking happens on the hypertable on the AN, because foreign tables will not have any triggers on them?

erimatnor replied:
Yes, and we assume that the chunks on DNs have the same triggers. This is the best we can do. This might warrant having a GUC to disable this plan node, but not sure it is necessary at this point (currently providing an anonymous session variable).

nikkhils commented on tsl/test/sql/dist_hypertable.sql.in:
```diff
@@ -359,18 +359,60 @@ SELECT * FROM disttable;
 RETURNING 1;
 
--- On conflict
+-- On conflict. Should use a DataNodeDispatch plan and NOT a
```
I believe there should be no impact on using COPY to insert into compressed chunks? Do we need a test case to cover inserting into a compressed chunk?

nikkhils replied:
Sven pointed out that COPY to compressed chunks does not work and I believe it's been tracked elsewhere. So resolving this here for now.

pmwkaa reviewed (commented):
Looks good overall.

I am thinking that there might be not much sense initially to associate `COPY` mode with batching from performance point of view, unless it is easy to implement. Because of the `COPY` command nature, we do not expect to receive reply for each row as far as I understand.

nikkhils commented on tsl/test/sql/dist_hypertable.sql.in:
```diff
@@ -810,6 +852,8 @@ ORDER BY time;
 $$);
 
+
+SET timescaledb.max_insert_batch_size=10;
```
I don't see the added tests erroring out when COPY fails. We could maybe drop an attribute on a DN and then INSERT into that remote chunk and see COPY error out.

erimatnor replied:
Added.

erimatnor commented on tsl/src/remote/connection.c:
```diff
@@ -1852,6 +1954,137 @@ remote_connection_set_single_row_mode(TSConnection *conn)
+
+	PQclear(res);
+
+	if (binary && !send_binary_copy_header(conn, err))
```
I am not 100% sure as I did not implement this (just moved and refactored the code).

But, I believe this might help explain: Basically, the `FORMAT BINARY` in the `COPY` statement has nothing to do with the wire protocol in `libpq`. It just tells the system the format of a "file" read by the server. Note that if you do the following statement:

```
COPY mytable FROM 'my_data_file` FORMAT BINARY;
```

you are basically telling the _server_ to open and read a file in binary format that must exist on the server. This is different from the `libpq` wire format when sending the data from the client to the server.

When we are sending data from a client, the COPY statement sets things up to read from STDIN instead of a file. In other words, the server does not read the file locally.

The copy header sent here is the format of the data sent with `libpq` from the client to the server and not file format.

pmwkaa replied:
I would rather reference this one instead https://www.postgresql.org/docs/current/protocol-flow.html#PROTOCOL-COPY. I was curious why we had to do this hack and send raw protocol header instead of using libpq, which likely would do the same under the hood

erimatnor commented:
@nikkhils @pmwkaa I refactored the tests. Mainly I wanted to keep the old tests as is instead of reworking them for the `DataNodeCopy` case. Instead, I have added tests related to `DataNodeCopy` to the end of the `dist_hypertable` test file. Should be easier to see the diff now.

nikkhils reviewed (approved):
LGTM now with the test changes. This is great work! 👍

erimatnor commented:
@nikkhils @pmwkaa Got some feedback from Sven, and I need to add checks for generated columns and default column values when using RETURNING clause.

pmwkaa reviewed (approved):
Looks great overall, hope I didn't miss anything

erimatnor commented:
@nikkhils and @pmwkaa I added tests for disabling DataNodeCopy with serial, generated, and default columns when used in combination with a RETURNING clause.

svenklemm commented on tsl/src/planner.c:
```diff
@@ -146,3 +149,82 @@ tsl_set_rel_pathlist(PlannerInfo *root, RelOptInfo *rel, Index rti, RangeTblEntr
+copy_mode_enabled(void)
+{
+	const char *enable_copy =
+		GetConfigOption("timescaledb.enable_distributed_insert_with_copy", true, false);
```
I think this should be real GUC, all the other planner switches are real GUCs and it would give proper datatype support. Most of those planner GUCs are not expected to be changed by users during normal operations but they are a useful tool for support in  case of bugs and having them be a real GUC makes them more convenient as they are discoverable.

erimatnor replied:
I think the value of a GUC is limited for the user and leaves us less flexibility to change the INSERT path in the future. This is not a `SELECT`/query path where you'd like to pick one node over the other depending on how you'd like to optimize your query. Instead, I'd prefer the option to continue to evolve and experiment with the INSERT path, potentially replacing this node with something else in the future (given various caveats, etc). If we expose a GUC, then that's an "official" settings variable that users might come to rely on and it's difficult to remove or change.

This can always be elevated to GUC in the future.

mkindahl commented on tsl/src/remote/connection.c:
```diff
@@ -170,6 +171,147 @@ typedef struct TSConnection
+		i++;
+	}
+
+	pg_unreachable();
```
Macro `pg_unreachable` is intended for cases where the full situation cannot be caught by the compiler, for example, after a statement that will transfer control elsewhere or that will never return, not really for when you know that the list of enumerations (for example) is covered. It typically triggers undefined behavior. I would suggest to remove this line and just return error, or remove this line and log an error with `ereport`.

erimatnor replied:
I disagree. It is there for two reasons:

1. To safe-guard that the code above doesn't miss an enumeration or to catch that if upstream adds a new enumeration. Note that the compiler cannot really know since the above is a string-comparison in a for loop, not, e.g., a switch.
2. To let the people reading the code know that the one who wrote the code intentionally designed the code to not reach this point. Otherwise, someone would have asked to add a test case for this point based on the code coverage complaint.

If code coverage was smart enough, the macro would also help it to understand that this part of the code is not reachable and then it wouldn't complain.

I don't think we should spend time formulating errors that will never be thrown.

erimatnor replied:
Btw, in PostgreSQL, they use `pg_unreachable` exactly for the case you are arguing against: https://github.com/postgres/postgres/blob/e6ccd1ce1644d1b40b7981f8bc172394de524f99/src/backend/tcop/dest.c#L158

mkindahl commented on tsl/src/remote/connection.c:
```diff
@@ -170,6 +171,147 @@ typedef struct TSConnection
+/*
+ * Fill a connection error based on the result of a remote query.
+ */
+static bool
+fill_result_error(TSConnectionError *err, int errcode, const char *errmsg, const PGresult *res)
```
This always returns `false`: you can remove the return type (or is the last `return false` a typo?)

erimatnor replied:
Yes, that's intentional since it allows returning from other functions in one-line instead of two, like so:

```
if (error)
   return fill_error(....);
```

instead of:

``` 
if (error)
{
      fill_error(...);
      return false;
)
```

If you feel strongly that we should always do the second option, I can change.

mkindahl replied:
No strong opinion, this is not important and you can leave it as it is, but piggybacking that functionality on the function seems to be unnecessary. To stick it on one line one could just as well write:
```
if (error)
  return fill_error(...), false;
```

mkindahl commented on tsl/src/remote/txn.c:
```diff
@@ -90,6 +90,7 @@ remote_txn_begin(RemoteTxn *entry, int curlevel)
 	{
 		const char *sql;
 
+		Assert(remote_connection_get_status(entry->conn) == CONN_IDLE);
```
If `remote_connection_get_status` has side-effects, that might hide bugs later in the code that surface once the assertion is removed.

erimatnor replied:
Why do you think `remote_connection_get_status` has side effects? It is just a simple "getter".

mkindahl replied:
It is not a problem in the current code, but it might be the result of a future change, e.g., reading the status code from a receive buffer.

erimatnor replied:
I am not sure I follow. And why is reading status from a buffer a side-effect (unless it also modifies the buffer)?

mkindahl reviewed (commented):
Mostly small comments and nits, but it would be good if we could avoid using macros. They have a tendency to hide problems. I also wonder why the user id was removed.

mfundul commented on tsl/src/remote/connection.h:
```diff
@@ -116,6 +144,68 @@ extern void remote_connection_stats_reset(void);
+ * remote command execution. They need to be macros to preserve the error
+ * context of where they are called (line number, statement, etc.).
+ */
+#define remote_connection_error_elog(err, elevel)                                                  \
+	ereport(elevel,                                                                                \
+			((err)->remote.errcode != 0 ? errcode((err)->remote.errcode) :                         \
+										  errcode((err)->errcode),                                 \
+			 (err)->remote.msg ?                                                                   \
+				 errmsg_internal("[%s]: %s", (err)->nodename, (err)->remote.msg) :                 \
+				 ((err)->connmsg ? errmsg_internal("[%s]: %s", (err)->nodename, (err)->connmsg) :  \
+								   errmsg_internal("[%s]: %s", (err)->nodename, (err)->msg)),      \
+			 (err)->remote.detail ? errdetail_internal("%s", (err)->remote.detail) : 0,            \
+			 (err)->remote.hint ? errhint("%s", (err)->remote.hint) : 0,                           \
+			 (err)->remote.sqlcmd ? errcontext("Remote SQL command: %s", (err)->remote.sqlcmd) :   \
+									0))
```
I find `static inline` functions in header files to be perfect for this. You still get the same overhead as a macro since you avoid the function call, but you still get full type checking.

mkindahl replied:
In general, I do not worry about function calls. They are quite performant on modern architectures.

mfundul replied:
I still see some value in avoiding them when possible, for various reasons that may be more relevant in IOT devices, e.g.:
- You take advantage of instruction cache locality since you don't have to move the CPU Instruction Pointer.
- You can save some CPU register space and more importantly stack memory space by avoiding the call.
- Virtualised environments (hypervisors) may magnify this kind of overheads.

But you're probably right that the function call overhead itself is kind of irrelevant.

erimatnor replied:
This needs to be a macro to capture the context of where the error is through, like line-number, function where error is thrown, etc. If it is not a macro, it believe it will look as if all errors are thrown from the same place.

mkindahl replied:
That is right. Ok, then we have to keep the macro.

erimatnor replied:
FWIW, there's actually a comment above the macro that explains why it needs to be a macro.

mfundul replied:
One way would be this https://stackoverflow.com/questions/54049362/tracing-static-inline-function but  it is probably not worth the effort

erimatnor commented on tsl/src/remote/connection.h:
```diff
@@ -80,22 +102,28 @@ typedef enum TSConnectionResult
 
+typedef enum TSConnectionStatus
+{
+	CONN_IDLE,		 /* No command being processed */
+	CONN_PROCESSING, /* Command/query is being processed */
+	CONN_COPY_IN,	/* Connection is in COPY_IN mode */
```
Using `CONN_` prefix is consistent with existing code, which already uses `CONN_`  for other enums, `CONN_OK`, etc. I'll leave this refactoring  for another PR.

erimatnor commented on tsl/src/remote/connection.h:
```diff
@@ -116,6 +144,68 @@ extern void remote_connection_stats_reset(void);
+ * elevel: error level to use (typically ERROR, but might be less)
+ * res: PGresult containing the error
+ */
+#define remote_result_elog(pgres, elevel)                                                          \
+	do                                                                                             \
+	{                                                                                              \
+		PG_TRY();                                                                                  \
+		{                                                                                          \
+			TSConnectionError err;                                                                 \
+			remote_connection_get_result_error(pgres, &err);                                       \
+			remote_connection_error_elog(&err, elevel);                                            \
+		}                                                                                          \
+		PG_CATCH();                                                                                \
+		{                                                                                          \
+			PQclear(pgres);                                                                        \
+			PG_RE_THROW();                                                                         \
+		}                                                                                          \
+		PG_END_TRY();                                                                              \
+	} while (0)
```
Same here about capturing error context.

erimatnor commented on tsl/src/remote/connection.h:
```diff
@@ -116,6 +144,68 @@ extern void remote_connection_stats_reset(void);
+		PG_END_TRY();                                                                              \
+	} while (0)
+
+#define remote_connection_elog(conn, elevel)                                                       \
+	do                                                                                             \
+	{                                                                                              \
+		TSConnectionError err;                                                                     \
+		remote_connection_get_error(conn, &err);                                                   \
+		remote_connection_error_elog(&err, elevel);                                                \
+	} while (0)
```
Again same as above.

erimatnor commented on tsl/src/planner.c:
```diff
@@ -146,3 +149,82 @@ tsl_set_rel_pathlist(PlannerInfo *root, RelOptInfo *rel, Index rti, RangeTblEntr
+		}
+	}
+
+	if (copy_possible)
+		return data_node_copy_path_create(root, mtpath, hypertable_rti, subplan_index);
+
+	return data_node_dispatch_path_create(root, mtpath, hypertable_rti, subplan_index);
```
The flag is not about cost, it is about compatibility/correctness. COPY is always preferred, but it is not compatible with some inserts with RETURNING clauses, so we need to fall back to the "slower", but more general path. 

I can't imagine why COPY would ever be slower than the other alternative. Even if picking based on cost would make sense, that requires a good costing model, which we don't have.

erimatnor commented on tsl/src/nodes/data_node_dispatch.c:
```diff
@@ -152,7 +153,6 @@ typedef struct DataNodeDispatchState
 	DispatchState prevstate; /* Previous state in state machine */
 	DispatchState state;	 /* Current state in state machine */
 	Relation rel;			 /* The (local) relation we're inserting into */
-	Oid userid;				 /* User performing INSERT */
```
It wasn't used. It's a legacy thing that didn't get removed in a previous change. The userid is also available in the ChunkInsertState.

erimatnor commented on tsl/src/planner.c:
```diff
@@ -146,3 +149,82 @@ tsl_set_rel_pathlist(PlannerInfo *root, RelOptInfo *rel, Index rti, RangeTblEntr
+	if (NULL == enable_copy)
+		return true;
+
+	return strcmp(enable_copy, "true") == 0;
```
This is not meant for user consumption and won't be documented so it needs not care about `1` or `yes`, as explained in the comment. It is used _only_ for testing (in our tests). It can be elevated to a GUC later, but I don't think we should expose GUCs if we don't need to since they are public interfaces that we need to document and maintain.

svenklemm replied:
If this is for testing only then this should probably be only active in debug builds. Your code still exposes a GUC here even if it's not registered as one. Also this was brought up by multiple reviewers now.

erimatnor commented:
I rebased after merging the support for distributed triggers, then moved some tests to `dist_triggers` that was introduced by the previous PR.
